tls/design considerations.md

```diff
@@ -0,0 +1,47 @@
+
+One optimisation that was considered when designing the protocol was to use the libp2p host key to generate the certificate in the case of RSA and ECDSA keys (which we can assume to be supported signature schemes by all peers). That would have allowed us to strip the host key and the signature from the key extension, in order to
+
+1. reduce the size of the certificate and
+2. reduce the number of signature verifications the peer has to perform from 2 to 1.
+
+The protocol does not include this optimisation, because
+
+1. assuming that the peer uses an ECDSA key for generating the self-signed certificate, this only saves about ~150 bytes if the host key is an ECDSA key as well, and it even slightly increases the size of the certificate in case of a RSA host key. Furthermore, for ECDSA keys, the size of all handshake messages combined is less than 900 bytes, so having a slightly larger certificate won't require us to send more (TCP / QUIC) packets.
+2. For a client, the number of signature verifications shouldn't pose a problem, since it controls the rate of its dials. Only for servers this might be a problem, since a malicious client could force a server to waste resources on signature verification. However, this is not a particularly interesting DoS vector, since the client's certificate is sent in its second flight (after receiving the ServerHello and the server's certificate), so it requires the attacker to actually perform most of the TLS handshake, including encrypting the certificate chain with a key that's tied to that handshake.
+
+
+### Versioning - How we could roll out a new version of this protocol in the future
+
+An earlier version of this document included a version negotiation mechanism. While it is a desireable property to be able to change things in the future, it also adds a lot of complexity.
+
+To keep things simple, the current proposal does not include a version negotiation mechanism. A future version of this protocol might:
+
+1. Change the format in which the keys are transmitted. A x509 extension has an ID (the Objected Identifier, OID), so we can use a new OID if we want to change the way we encode information. x509 certificates allow use to include multiple extensions, so we can even send the old and the new version during a transition period. In the handshake protocol defined here, peers are required to skip over extensions that they don't understand.
+2. For more involved changes, a new version might (ab)use the SNI field in the ClientHello to announce support for new versions. To allow for this to work, the current version requires clients to send anything in the SNI field and server to completely ignore this field, no matter what its contents are.
```

Comment:
What was our motivation for using SNI instead of a custom extension again?

Reply:
A custom extension would definitely be the cleaner way to implement this. However, we‘d have to fork the Go standard library implementation and add support for that ourselves. Not sure how this would work out in JS- and Rust-land.

Reply:
Given that we don't even _need_ this yet, we can probably cross that bridge when we get there. Honestly, I'd kind of like to be able to use the SNI field for the expected server peer ID (for multiplexing multiple peers through some kind of reverse proxy).